src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -235,4 +249,104 @@ theorem sle_eq_carry (x y : BitVec w) :
     x.sle y = !((x.msb == y.msb).xor (carry w y (~~~x) true)) := by
   rw [sle_eq_not_slt, slt_eq_not_carry, beq_comm]
 
+/-! ### mul recurrence for bitblasting -/
+
+def mulRec (l r : BitVec w) (s : Nat) : BitVec w :=
+  let cur := if r.getLsb s then (l <<< s) else 0
+  match s with
+  | 0 => cur
+  | s + 1 => mulRec l r s + cur
+
+theorem mulRec_zero_eq (l r : BitVec w) :
+    mulRec l r 0 = if r.getLsb 0 then l else 0 := by
+  simp [mulRec]
+
+theorem mulRec_succ_eq (l r : BitVec w) (s : Nat) :
+    mulRec l r (s + 1) = mulRec l r s + if r.getLsb (s + 1) then (l <<< (s + 1)) else 0 := by
+  simp [mulRec]
+
+theorem zeroExtend_truncate_succ_eq_zeroExtend_truncate_of_getLsb_false
```

Comment:
Can all these `zeroExtend_truncate_*` lemmas live in `Lemmas`?

Reply:
I thought about this --- they need lemmas about `or` and `and`, which occur after `zeroExtend` in the `Lemmas` file. I could add them to the bottom of the `Lemmas` file if you prefer.

Reply:
Yes please. 
```
/-! ### zeroExtend, truncate, and bitwise operations -/
```

Reply:
Ah, in attempting the move, I realised that this also depends on  `add_eq_or_of_and_eq_zero`, which uses `add_eq_adc` that currently lives in `Bitblast.lean`, so this cannot be moved to `Lemmas.lean`, unfortunately.

Reply:
Of course, no problem then.

Reply:
But some of the lemmas *below* can move, right?

Reply:
Ah, right! Moved.